File: src/routes/yield-vault.js
/**
 * HiveBank Yield Vault — Routes
 * ==============================
 * POST /v1/bank/vault/deposit      — deposit USDC, create vault if new
 * GET  /v1/bank/vault/rates        — live APY from all 4 protocols (PUBLIC)
 * GET  /v1/bank/vault/stats        — TVL, yield, rebalance count (PUBLIC)
 * POST /v1/bank/vault/rebalance    — manual rebalance trigger (x-hive-internal required)
 * GET  /v1/bank/vault/:did         — vault balance, yield, allocation
 * POST /v1/bank/vault/withdraw     — burn shares, return USDC + yield
 *
 * Auth:
 *   deposit / withdraw  — did in body, fully public (no key required)
 *   rebalance trigger   — x-hive-internal header required
 *   rates / stats       — fully public
 */

'use strict';

const express = require('express');
const router  = express.Router();
const yieldVault = require('../services/yield-vault');

const INTERNAL_KEY = process.env.HIVE_INTERNAL_KEY ||
  'hive_internal_125e04e071e8829be631ea0216dd4a0c9b707975fcecaf8c62c6a2ab43327d46';

// ─────────────────────────────────────────────────────────────────────────────
//  PUBLIC ENDPOINTS (no auth)
// ─────────────────────────────────────────────────────────────────────────────

// GET /v1/bank/vault/rates — current APY from all 4 protocols + best protocol
router.get('/rates', async (req, res) => {
  try {
    const rates = await yieldVault.getRates();
    res.json({ success: true, data: rates });
  } catch (err) {
    res.status(500).json({ success: false, error: 'Failed to fetch rates', detail: err.message });
  }
});

// GET /v1/bank/vault/stats — total TVL, yield earned, rebalance count
router.get('/stats', async (req, res) => {
  try {
    const stats = await yieldVault.getStats();
    res.json({ success: true, data: stats });
  } catch (err) {
    res.status(500).json({ success: false, error: 'Failed to fetch stats', detail: err.message });
  }
});

// ─────────────────────────────────────────────────────────────────────────────
//  DEPOSIT — public (did in body is the identity check)
// ─────────────────────────────────────────────────────────────────────────────

// POST /v1/bank/vault/deposit — body: { did, amount_usdc }
router.post('/deposit', async (req, res) => {
  const { did, amount_usdc } = req.body;

  if (!did) {
    return res.status(400).json({
      success: false,
      error: 'did is required',
      hint: 'Provide your Hive DID: e.g. "did:hive:your-agent-id"',
    });
  }
  if (amount_usdc === undefined || amount_usdc === null) {
    return res.status(400).json({ success: false, error: 'amount_usdc is required' });
  }
  if (isNaN(Number(amount_usdc)) || Number(amount_usdc) <= 0) {
    return res.status(400).json({ success: false, error: 'amount_usdc must be a positive number' });
  }

  try {
    const result = await yieldVault.deposit(did, amount_usdc);
    if (result.error) return res.status(400).json({ success: false, error: result.error });
    res.status(201).json(result);
  } catch (err) {
    res.status(500).json({ success: false, error: 'Deposit failed', detail: err.message });
  }
});

// ─────────────────────────────────────────────────────────────────────────────
//  WITHDRAW — public (did in body)
// ─────────────────────────────────────────────────────────────────────────────

// POST /v1/bank/vault/withdraw — body: { did, amount_usdc }
router.post('/withdraw', async (req, res) => {
  const { did, amount_usdc } = req.body;

  if (!did) {
    return res.status(400).json({ success: false, error: 'did is required' });
  }
  if (amount_usdc === undefined || amount_usdc === null) {
    return res.status(400).json({ success: false, error: 'amount_usdc is required' });
  }
  if (isNaN(Number(amount_usdc)) || Number(amount_usdc) <= 0) {
    return res.status(400).json({ success: false, error: 'amount_usdc must be a positive number' });
  }

  try {
    const result = await yieldVault.withdraw(did, amount_usdc);
    if (result.error) return res.status(400).json({ success: false, error: result.error, data: result });
    res.json(result);
  } catch (err) {
    res.status(500).json({ success: false, error: 'Withdrawal failed', detail: err.message });
  }
});

// ─────────────────────────────────────────────────────────────────────────────
//  MANUAL REBALANCE — requires x-hive-internal header
// ─────────────────────────────────────────────────────────────────────────────

// POST /v1/bank/vault/rebalance
router.post('/rebalance', async (req, res) => {
  const key = req.headers['x-hive-internal'];
  if (!key || key !== INTERNAL_KEY) {
    return res.status(401).json({
      success: false,
      error: 'INTERNAL_AUTH_REQUIRED',
      hint: 'Provide x-hive-internal header with the internal key',
    });
  }

  try {
    const result = await yieldVault.manualRebalance();
    res.json(result);
  } catch (err) {
    res.status(500).json({ success: false, error: 'Rebalance failed', detail: err.message });
  }
});

// ─────────────────────────────────────────────────────────────────────────────
//  GET VAULT BY DID — public
// ─────────────────────────────────────────────────────────────────────────────

// GET /v1/bank/vault/:did — vault balance, current yield, protocol allocation
// NOTE: must be defined AFTER /rates, /stats, /deposit, /withdraw, /rebalance
//       to avoid those static path segments being matched as :did
router.get('/:did(*)', async (req, res) => {
  const { did } = req.params;

  if (!did) {
    return res.status(400).json({ success: false, error: 'did is required in path' });
  }

  try {
    const result = await yieldVault.getVaultBalance(did);
    if (result.error) return res.status(404).json({ success: false, error: result.error });
    res.json({ success: true, data: result });
  } catch (err) {
    res.status(500).json({ success: false, error: 'Failed to fetch vault', detail: err.message });
  }
});

module.exports = router;

File: yield_vault.py
"""
HiveBank Yield Vault
====================
Set and Forget USDC vault. Deposit once, earn maximum DeFi yield automatically.

Phase 1 (now): Paper trading — tracks simulated positions, real APY data, real P&L math.
              No on-chain transactions yet. Proves the model before deploying capital.
Phase 2 (when Steve has $50K+): Switch SIMULATED=False, wire Coinbase CDP for real txns.

All APY data is real (fetched live). All math is real. Only the execution is simulated.

─────────────────────────────────────────────────────────────────────────────────
NOTE: This Python file is the architecture specification / reference document.
      The live implementation is in Node.js (Express) to match the existing
      HiveBank service:
        src/services/yield-vault.js  — YieldMonitor, NAVUpdater, Rebalancer, DB logic
        src/routes/yield-vault.js    — Express routes: /v1/bank/vault/*
      This Python file documents the design contract for Phase 2 CDP integration.
─────────────────────────────────────────────────────────────────────────────────

## Toggle
SIMULATED = True   # Phase 1: paper trading, real APY feeds, no on-chain txns
                   # Phase 2: set False → execution layer activates via Coinbase CDP

## Architecture

Vault             — holds deposited USDC, tracks shares per depositor
YieldMonitor      — background thread, polls APY from protocols every 15 min
NAVUpdater        — background thread, accrues yield to all vaults every 15 min
Rebalancer        — background thread, executes rebalance when spread > threshold
Protocol adapters — read-only APY fetchers for Aave, Morpho, Spark, Compound on Base

## Protocol APY Sources

1. Aave V3 Base:
   GET https://aave-api-v2.aave.com/data/markets-data/0x2f39d218133AFaB8F2B819B1066c7E434Ad94E9e
   Fallback: 2.5%

2. Morpho Blue Base:
   POST https://blue-api.morpho.org/graphql
   Body: {"query": "{ markets(where: {collateralAsset: {symbol_in: [\"USDC\"]}, whitelisted: true}, orderBy: supplyApy, orderDirection: desc, first: 5) { items { uniqueKey supplyApy collateralAsset { symbol } loanAsset { symbol } } } }"}
   Take highest USDC supply APY.
   Fallback: 5.5%

3. Spark Protocol:
   GET https://api.spark.fi/v1/rates
   Fallback: 4.2%

4. Compound V3 Base:
   GET https://api.compound.finance/api/v2/ctoken?addresses[]=0x46e6b214b524310239732D51387075E0e70970bf
   Fallback: 3.8%

## Vault Logic

Deposit: User deposits N USDC → shares = N / nav_per_share
Rebalance trigger: best_apy - current_apy > 0.5% AND simulated_gas < 1 day yield improvement
NAV update (every 15 min):
    elapsed_days = (now - last_rebalance) / 86400
    daily_yield = (current_apy / 365) * total_deposits
    nav_per_share += daily_yield / total_shares
Withdrawal: shares burned → USDC = shares * nav_per_share

## Database Schema (PostgreSQL)

CREATE TABLE IF NOT EXISTS yield_vaults (
    vault_id TEXT PRIMARY KEY,
    owner_did TEXT NOT NULL,
    deposited_usdc NUMERIC(18,6) NOT NULL DEFAULT 0,
    shares NUMERIC(18,6) NOT NULL DEFAULT 0,
    created_at TIMESTAMPTZ DEFAULT NOW(),
    last_activity TIMESTAMPTZ DEFAULT NOW()
);

CREATE TABLE IF NOT EXISTS vault_allocations (
    id SERIAL PRIMARY KEY,
    protocol TEXT NOT NULL,
    allocated_usdc NUMERIC(18,6) NOT NULL DEFAULT 0,
    current_apy NUMERIC(8,4) NOT NULL DEFAULT 0,
    last_rebalanced TIMESTAMPTZ DEFAULT NOW(),
    is_active BOOLEAN DEFAULT TRUE
);

CREATE TABLE IF NOT EXISTS vault_events (
    id SERIAL PRIMARY KEY,
    vault_id TEXT NOT NULL,
    event_type TEXT NOT NULL, -- deposit, withdraw, rebalance, yield_accrual
    amount_usdc NUMERIC(18,6),
    protocol_from TEXT,
    protocol_to TEXT,
    apy_at_time NUMERIC(8,4),
    simulated BOOLEAN DEFAULT TRUE,
    timestamp TIMESTAMPTZ DEFAULT NOW()
);

## Routes

POST /v1/bank/vault/deposit       — {did, amount_usdc} — create vault if new, add shares
GET  /v1/bank/vault/:did          — balance, yield, allocation, projected APY
POST /v1/bank/vault/withdraw      — {did, amount_usdc} — burn shares, return principal + yield
GET  /v1/bank/vault/rates         — current APY from all 4 protocols (PUBLIC)
GET  /v1/bank/vault/stats         — TVL, yield earned, rebalance count (PUBLIC)
POST /v1/bank/vault/rebalance     — manual trigger (x-hive-internal header required)

## Auth

deposit/withdraw:  did in body — public
rebalance trigger: x-hive-internal must equal $HIVE_INTERNAL_KEY env (rotated 2026-04-25; prior key DEAD)
rates / stats:     fully public

## Phase 2 Activation (Coinbase CDP)

When SIMULATED = False:
  - Wire Coinbase CDP SDK for on-chain USDC transfers
  - Use DATABASE_URL (os.environ.get('DATABASE_URL')) — never hardcoded
  - Deploy real capital: minimum $50K recommended for gas efficiency
  - Expected APY uplift vs. static vaults: +1.5–3% through protocol rotation

"""

# ─── Phase toggle — flip to False when capital is ready ───────────────────────
SIMULATED = True

# ─── DO NOT hardcode DATABASE_URL — always use os.environ.get('DATABASE_URL') ─
import os
DATABASE_URL = os.environ.get('DATABASE_URL')  # Set as Render env var

# Phase 2: import coinbase_agentkit or CDP SDK here when SIMULATED = False
